tests/src/androidTest/java/com/squareup/duktape/DuktapeSetTest.java

```diff
@@ -494,4 +497,191 @@ public String getValue() {
           .hasMessage("TypeError: boolean required, found 'ninja' (stack index -1)");
     }
   }
+
+  interface ObjectSorter {
+    Object[] sort(Object[] args);
+  }
+
+  @Test public void arraysOfObjects() {
+    duktape.set("Sorter", ObjectSorter.class, new ObjectSorter() {
+      @Override
+      public Object[] sort(Object[] args) {
+        if (args == null) return null;
+        Arrays.sort(args, new Comparator<Object>() {
+          @Override
+          public int compare(Object l, Object r) {
+            if (l == null) return -1;
+            if (r == null) return 1;
+            return ((Comparable<Object>) l).compareTo(r);
+          }
+        });
+        return args;
+      }
+    });
+
+    assertThat(duktape.evaluate("Sorter.sort(null)")).isNull();
+    assertThat((Object[]) duktape.evaluate("Sorter.sort([2, 4, 3, 1])"))
+        .isEqualTo(new Object[]{1.0, 2.0, 3.0, 4.0});
+
+    assertThat((Object[]) duktape.evaluate("Sorter.sort(['b', 'd', null, 'a'])"))
+        .isEqualTo(new String[]{null, "a", "b", "d"});
+
+    TimeZone original = TimeZone.getDefault();
+    try {
+      TimeZone.setDefault(TimeZone.getTimeZone("GMT+0:00"));
+      duktape.evaluate("Sorter.sort([ 1, 2, 3, new Date(0) ])");
+      fail();
+    } catch (DuktapeException expected) {
+      assertThat(expected)
+          .hasMessage("Error: Cannot marshal return value 1970-01-01 00:00:00.000+00:00 to Java");
+    } finally {
+      TimeZone.setDefault(original);
+    }
+  }
+
+  interface StringSorter {
+    String[] sort(String[] args);
+  }
+
+  @Test public void arraysOfStrings() {
+    duktape.set("Sorter", StringSorter.class, new StringSorter() {
+      @Override public String[] sort(String[] args) {
+        Arrays.sort(args);
+        return args;
+      }
+    });
+
+    assertThat((Object[]) duktape.evaluate("Sorter.sort(['b', 'd', 'c', 'a'])"))
+        .isEqualTo(new String[]{ "a", "b", "c", "d" });
+
+    try {
+      duktape.evaluate("Sorter.sort(['b', 'd', 3, 'a'])");
+      fail();
+    } catch (DuktapeException expected) {
+      assertThat(expected).hasMessage("TypeError: string required, found 3 (stack index -1)");
```

Comment:
`stack index -1` is leaking implementation details. Will probably change if you build another mechanism for types